### ISSUE TITLE
Equivalency fallback for non-IComparable types can leave CollectionTally in corrupt state

### ISSUE THREAD

**Issue by slabarque** [nunit/nunit#3976]
When checking 2 arrays for equivalency it should not be required for the object to be comparable, equatable should be enough.
Take a look at following tests that demonstrate the issue. The last test should not have failed.

```
        [Test]
        public void ObjectArrayEquivalency()
        {
            var message3 = new object();
            var message4 = new object();
            var array = new[]
            {
                message3,
                message4
            };

            //Succeeds
            Assert.That(array, Is.EquivalentTo(new[]
            {
                message4,
                message3
            }));
        }

        [Test]
        public void TuplesWithComparableArrayEquivalency()
        {
            var message3 = DateTime.Now;
            var message4 = DateTime.Now.AddDays(1);
            var array = new[]
            {
                new Tuple<int, object, CancellationToken>(1, message3, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(1, message4, CancellationToken.None)
            };

            //Succeeds
            Assert.That(array, Is.EquivalentTo(new[]
            {
                new Tuple<int, object, CancellationToken>(1, message4, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(1, message3, CancellationToken.None)
            }));
        }



        [Test]
        public void TuplesWithObjectArrayEquivalency()
        {
            var message3 = new object();
            var message4 = new object();
            var array = new[]
            {
                new Tuple<int, object, CancellationToken>(1, message3, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(1, message4, CancellationToken.None)
            };

            //Fails on System.Array.Sort: Failed to compare two elements in the array. At least one object must implement IComparable.
            Assert.That(array, Is.EquivalentTo(new[]
            {
                new Tuple<int, object, CancellationToken>(1, message4, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(1, message3, CancellationToken.None)
            }));
        }
```

I'm using the following nuget package versions:
```
<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
<PackageReference Include="NUnit" Version="3.13.2" />
<PackageReference Include="NUnit3TestAdapter" Version="4.0.0" />
```

When bumping the versions a bit lower to earlier versions (where I knew my test did not fail) the problem is not there.
These versions do not have this issue:
```
<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
<PackageReference Include="NUnit" Version="3.11.0" />
<PackageReference Include="NUnit3TestAdapter" Version="3.13.0" />
```

**Comment by mikkelbu** [comment]
I think this is a duplicate of #3841 which is fixed on master, @slabarque can you test if the problem is solved if you use the nuget package from https://ci.appveyor.com/project/CharliePoole/nunit/builds/41343183/artifacts

**Comment by slabarque** [comment]
Nice, that version indeed fixes the problem. Any idea when this will be available as stable release on Nuget.org?

**Comment by slabarque** [comment]
Just tested the new package version on my actual code but discovered there is still an issue with it.
The following test should not fail:
```
        [Test]
        public void TuplesWithObjectArrayEquivalency()
        {
            var message3 = new object();
            var message4 = new object();
            var array = new[]
            {
                new Tuple<int, object, CancellationToken>(1, message3, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(2, message3, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(1, message4, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(2, message4, CancellationToken.None)
            };

            //Fails on System.Array.Sort: Failed to compare two elements in the array. At least one object must implement IComparable.
            Assert.That(array, Is.EquivalentTo(new[]
            {
                new Tuple<int, object, CancellationToken>(1, message4, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(2, message4, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(1, message3, CancellationToken.None),
                new Tuple<int, object, CancellationToken>(2, message3, CancellationToken.None)
            }));
        }
```

**Comment by stevenaw** [comment]
Thanks for following up on this @slabarque 
I can confirm this is an issue on `v3.13-dev`. It looks like when NUnit attempts to sort a tuple when some of the tuple's members don't implement icomparable, it mutates the collection and will not revert it if an exception is thrown. With the example code above (on netfx 4.5) the exception is thrown after one half of a element swap has been performed, which is corrupting the arraylist.

The error is happening here:
https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/CollectionTally.cs#L90

**Comment by mikkelbu** [comment]
Closed by #3978
